src/core/thread/mle_router.cpp

```diff
@@ -3530,6 +3530,19 @@ ThreadError MleRouter::GetNextNeighborInfo(otNeighborInfoIterator &aIterator, ot
     return error;
 }
 
+void MleRouter::ResolveRoutingLoops(uint16_t aSourceMac, uint16_t aDestRloc16)
+{
+    if (aSourceMac == GetNextHop(aDestRloc16))
+    {
+        // loop detected
+        Router *router = GetRouter(GetRouterId(aDestRloc16));
+        assert(router != NULL);
+
+        // invalidate next hop
+        router->mNextHop = kInvalidRouterId;
```

Comment:
I might be wrong but I believe we should reset the trickle timer here as well, as 3rd paragraph of 5.9.7 describes.